ulib/FStar.List.Tot.Properties.fst

```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
     | [] -> ()
     | h::t -> map_lemma f t
 
+(** Properties about unsnoc *)
+
+(** [unsnoc] is the inverse of [snoc] *)
+val lemma_unsnoc_snoc: #a:Type -> l:list a{length l > 0} ->
+  Lemma (requires True)
+    (ensures (let l', x = unsnoc l in snoc l' x == l))
+    [SMTPat (unsnoc l)]
```

Comment:
This SMT pattern is worrisome. It fires on `unsnoc l` but contributes a term that contains `unsnoc l` as a strict sub term. This is risky as it may interact poorly with other patterns. Even aside from that, this pattern provides an equation about snoc/unsnoc in the context as soon as an unsnoc appears, which may be unnecessary. 

I suggest either

1. picking the pattern to be `snoc (fst (unsnoc l)) (snd (unsnoc l))`

2. removing the pattern and relying on explicit invocations at callsites.

Reply:
I'm going for the 1st option you suggested.

Reply:
From my testing, it looks like the pattern `snoc (fst (unsnoc l)) (snd (unsnoc l))` is too restrictive, and having an equation about snoc/unsnoc in context seems extremely useful (at least from my use cases, almost any time you want to use snoc, you also want unsnoc, and vice-versa).

I completely agree that purely `unsnoc l` is worrisome too, although I am not sure I understand the "as a strict subterm" part. Could you please clarify? Thanks :)

Reply:
I generally like to see patterns on quantifiers where the terms that result from the quantifier instantiation are smaller than the term used to instantiate the pattern. This is not a strict rule, but with this discipline, it is usually easier to argue that there are no matching loops, since the new terms produces after a quantifier instantiation is smaller than the pattern you started with.

Reply:
One possibility may be to define (uncurried) `snoc (l, x) = append l [x]` and then this lemma could be `snoc (unsnoc l) == l` with `snoc (unsnoc l)` as the pattern. I can see that `snoc (fst (unsnoc l)) (snd (unsnoc l))` could be too restrictive (since the `fst` and `snd` may not appear explcitly).

Reply:
Defining it with the uncurried version makes a lot of sense. In fact, I am pretty sure this is a better/cleaner definition. Should I also add in `lemma_snoc_unsnoc` (which says `unsnoc (snoc (l, x)) == (l, x)`) as well? I personally don't need it, but it seems like something that someone else using snoc/unsnoc might want/need.